### ISSUE TITLE
Maintaining consumer states across multiple rest proxy instances

### ISSUE THREAD

**Issue by Sharad001** [confluentinc/kafka-rest#212]
Hi all,

I have a problem in maintaining consumer instances across different Rest-Proxy instances because for that I have to deploy some infrastructure like "zookeeper" to maintain the states of the consumer across different proxy instances but as written on confluent site that it support distributed processing the above argument defies that . So I just want to know that confluent will provide this kind of functionality or not in their new versions because as of now if some rest instance dies than the client has to request for the new consumer again and it has to maintain the state of the consumer to start from that offset and I think this defies the underlying principle of Rest Client . So someone please see about this issue.

**Comment by ewencp** [comment]
The REST proxy uses normal Java consumers under the hood. Whether you store offsets in zookeeper or within Kafka (both of which you must have deployed to run REST proxy anyway), consumers will load any committed offsets. When a consumer instance is killed (or dies, e.g. because you shut down the proxy) the partitions are rebalanced across remaining consumer instances in the group and each loads the respective offsets.

Note that the proxy *does* disable auto offset commit. This is to allow you to ensure at least once semantics given the current implementation -- if auto offset commit was enabled, some data would be returned by a request and then the consumer could immediately do auto commit, even though your client wasn't done processing the data yet. Therefore, you need to periodically invoke [offset commit](http://docs.confluent.io/3.0.0/kafka-rest/docs/api.html#post--consumers-(string-group_name)-instances-(string-instance)-offsets) from your client application or provide the `auto.commit.enable` flag when [creating the consumer instance](http://docs.confluent.io/3.0.0/kafka-rest/docs/api.html#post--consumers-(string-group_name)) or offsets will never be committed and after rebalance consumers will always start whereever they automatically reset to (e.g. the beginning of the topic).

**Comment by Sharad001** [comment]
Hi sir,

Actually my doubt was that because my consumer is not running on client side it is running in proxy and I am just querying to the proxy to get the message so when proxy dies my consumer will also die and then when i send the request the load balancer will redirect me to some other proxy instance which is running but on that instance my consumer is not running so I should have to launch the consumer again but the request is for consuming message not launching consumer so how should I tackle this ?

**Comment by ewencp** [comment]
@Sharad001 You can use a load balancer, but need to [configure instances correctly](http://docs.confluent.io/3.0.0/kafka-rest/docs/deployment.html#kafkarest-deployment) and after creating a consumer you should use the [base URL](http://docs.confluent.io/3.0.0/kafka-rest/docs/api.html#post--consumers-(string-group_name)) returned in the initial request to create it. Subsequent reads will either fail (if the instance crashed and is not brought back up) or [reads will return HTTP status 404 and code 40403](http://docs.confluent.io/3.0.0/kafka-rest/docs/api.html#get--consumers-(string-group_name)-instances-(string-instance)-topics-(string-topic_name)) if the consumer instance is not found (which it won't be when the proxy process is revived).

No matter what you do next, the other consumer instances in the group will take over the partitions that the dead consumer instance used to own (i.e. the group rebalances). Whatever client was consuming messages can detect the two types of failure (either it can't contact the proxy or sees the 40403 error) and basically just reset itself, creating a new consumer instance (again, on any instance the load balancer directs it to) and then starting to consume messages again.

**Comment by Sharad001** [comment]
Hi sir,
I got it now so basically there is no mechanism as of now in kafka-rest  so that if consumer fails on some proxy then it will be instantiated by its own on some other proxy . So I think this should be provided as a feature because as of now client has to handle the burden of instantiating the consumer again.

**Comment by ewencp** [comment]
@Sharad001 We actually wouldn't want to add that functionality. It sounds convenient, but if the REST proxies recreated consumers that died automatically, clients wouldn't know *which* REST proxy to look for the consumer on. Even if we provided some lookup for that, you still have the problem that no instance of your application that even knows about the consumer instance may be alive. Which means nothing will be consuming data from it and all partitions assigned to that instance will not have any of their data processed.

In short, it is important that you *explicitly* create consumer instances from applications that are actually going to subsequently make requests to consume data. Otherwise your application may leave some partitions not making any progress for indefinite periods of time.

**Comment by Sharad001** [comment]
Thanks a lot @ewencp now I understand the whole point.
